demo/src/main.rs

```diff
@@ -0,0 +1,116 @@
+    .await;
+    let agent_faber = DemoAgent::new(
+        "faber".into(),
+        mediator_url.clone(),
+        mpsc_registry.clone(),
+        COLOR_GREEN.into(),
+    )
+    .await;
+
+    (agent_alice, agent_faber)
+}
+
+async fn workflow_connection_by_role(alice: DemoAgent, faber: DemoAgent) {
+    // This simulates trusted communication channel, over which Faber send Alice an invitation
+    // In real setting, the equivalent could be scanning a QR code on trustworthy https:// website of Faber company
+    let (channel_sender, channel_receiver) = mpsc::channel::<Invitation>(1);
+
+    // We start both workflow in separate threads, to simulate reality - both parties are typically different machines,
+    // often invitee being a mobile device, inviter often being a company server
+    // The 2 parties communicate together via didcomm, over network. The DemoAgent has built-in integration
+    // with HTTP server to receive didcomm messages.
+    let handle1 = tokio::task::spawn(async move { workflow_invitee(alice, channel_receiver).await });
+    let handle2 = tokio::task::spawn(async move { workflow_inviter(faber, channel_sender).await });
+    tokio::join!(handle1, handle2);
+}
+
+async fn workflow_invitee(alice: DemoAgent, mut rx_invitation: Receiver<Invitation>) {
+    let msg_oob_invitation = rx_invitation.recv().await.unwrap();
+    info!("****** Alice receives connection invitation via a trusted channel ******");
+    let invitee_requested = alice.reply_invitation(msg_oob_invitation).await;
+
+    let response = alice.wait_for_message::<Response>().await.unwrap();
+    info!("****** Alice receives 'connection-response', will reply back with 'connection-response' ******");
+    let invitee_complete = alice.reply_connection_response(invitee_requested, response).await;
+
+    info!("****** Alice and Faber can now securely exchange other messages via established didcomm connection ******");
+    alice
+        .send_message(invitee_complete, "Hello Faber, this is Alice.".into())
+        .await;
+
+    let msg_any = alice.wait_for_message::<AriesMessage>().await.unwrap();
+    info!("Alice received message {msg_any:?}");
+}
+
+async fn workflow_inviter(faber: DemoAgent, tx_invitation: Sender<Invitation>) {
+    info!("****** Faber is preparing connection invitation to pass to Alice ****** ");
+    let (msg_oob_invitation, faber_invite_info) = faber.prepare_invitation().await;
+    tx_invitation.send(msg_oob_invitation).await.unwrap();
+
+    info!("****** Faber received connection invitation via trusted means, will reply sending 'connection-request' message ******");
+    let msg_request = faber.wait_for_message::<Request>().await.unwrap();
+    // todo: make the decryption part of agent's receive function
+    let inviter_requested = faber.reply_connection_request(msg_request, faber_invite_info).await;
+
+    let msg_any = faber.wait_for_message::<AriesMessage>().await.unwrap();
+    info!("Faber received message {msg_any:?}");
+
```

Comment:
should we show that this is where faber would call to `acknowledge_connection` in order to transition into `InviterConnection<Completed>`?

looks a bit weird that `send_message_2` takes in an inviter who is still in the `Requested` state

Reply:
After refactoring, things are bit different now; message sending happens over the memory only and that been pointed out doesn't exist anymore.